src/Elmish/HTML/Internal.purs

```diff
@@ -1,9 +1,90 @@
+-- |
 foreign import data CSS :: Type
+instance jsCSS :: CanPassToJavaScript CSS
+
+
+-- | Construct a value of type `CSS`, which is the type of `style` props, out of
+-- | a record with all `String` values. For example:
+-- |
+-- |    div { style: css { height: "50px", width: "50px" } }
+-- |
+-- | There is currently no type safety regarding the specific fields admitted in
+-- | the style, or different types of those fields. This has been deemed "good
+-- | enough" for now.
+css :: forall r. Homogeneous r String => { | r } -> CSS
+css = unsafeCoerce
+
 
+-- | Retype a string as a React component, which is something that React
+-- | supports out of the box - i.e. a string "div" can be used in place of a
+-- | `React.Component`-descendant class
 unsafeCreateDOMComponent :: String -> ImportedReactComponent
 unsafeCreateDOMComponent = unsafeCoerce
+
+
+-- | A CSS-framework-friendly tag-creating function with props. The `reqProps`
+-- | type parameter is a row of optional props for this tag. The point of
+-- | CSS-framework-friendliness is that the `className` prop is given as first
+-- | parameter rather than being part of the other props, so we can write `div_
+-- | "row" {}` rather than `div { className: "row" }`
+type StyledTag_ optProps =
+  forall props content
+   . ReactChildren content
+  => IsSubsetOf props (optProps + CommonProps)
+  => ValidReactProps { | props }
+  => String -> { | props } -> content -> ReactElement
+
+-- | A CSS-framework-friendly tag-creating function without props. The point of
+-- | CSS-framework-friendliness is that the `className` prop is given as first
+-- | parameter rather than being part of the other props, so we can write `div
+-- | "row"` rather than `div { className: "row" }`
+type StyledTag =
+  forall content. ReactChildren content => String -> content -> ReactElement
+
+-- | See comments on `StyledTag`
+type StyledTagNoContent = String -> ReactElement
+
+-- | See comments on `StyledTag_`
+type StyledTagNoContent_ optProps =
+  forall props
+   . IsSubsetOf props (optProps + CommonProps)
+  => ValidReactProps { | props }
+  => String -> { | props } -> ReactElement
+
+styledTag_ :: forall props content
+   . ReactChildren content
+  => ValidReactProps props
+  => String -> String -> props -> content -> ReactElement
+styledTag_ tag cls props =
+  createElement (unsafeCreateDOMComponent tag) $ (unsafeCoerce props :: {}) `merge` { className: cls }
+
+styledTag :: String -> StyledTag
+styledTag tag cls content =
+  createElement (unsafeCreateDOMComponent tag) { className: cls } content
+
+styledTagNoContent :: String -> StyledTagNoContent
+styledTagNoContent tag cls =
+  createElement' (unsafeCreateDOMComponent tag) { className: cls }
+
+styledTagNoContent_ :: forall props
+   . ValidReactProps props
+  => String -> String -> props -> ReactElement
+styledTagNoContent_ tag cls props =
+  createElement' (unsafeCreateDOMComponent tag) $ (unsafeCoerce props :: {}) `merge` { className: cls }
```

Comment:
A tiny part of me wonders whether this should be a separate library to reduce bundle file size. I know tree shaking should solve that but it’s still not working reliably in the PS ecosystem.

Reply:
One has to balance the gain from splitting the libraries against the loss of managing too many libraries, and code duplication between them.

If this was a separate library, it would still have to have `elmish-html` as a dependency for the props types. Or it would have to duplicate them, as well as their generation code. And now you're maintaining two copies of that. And so on...

So after a few rounds of this thinking, I decided to not split.

Reply:
Thanks, that’s fair. I agree, management overhead can get big. Hopefully they make headway on https://github.com/purescript/purescript/pull/3791